File: android/kotlin/SoftEtherMobileClient.kt
package com.softether.mobile

import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

/**
 * SoftEther Mobile Client - Kotlin Bridge for Android
 * 
 * Type-safe Kotlin API wrapping the mobile FFI (libsoftether_mobile.so)
 * 
 * Architecture:
 *   Android VpnService (Kotlin) → SoftEtherMobileClient (this file) →
 *   mobile_jni.c (JNI wrapper) → libsoftether_mobile.so (mobile FFI) →
 *   Zig Packet Adapter
 */

// ============================================================================
// Data Classes
// ============================================================================

/**
 * VPN connection configuration
 */
data class VpnConfig(
    val serverName: String,
    val serverPort: Int = 443,
    val hubName: String = "VPN",
    val username: String,
    val password: String,
    
    // Connection options
    val useEncrypt: Boolean = true,
    val useCompress: Boolean = true,
    val halfConnection: Boolean = false,
    val maxConnection: Int = 1,
    
    // Performance tuning
    val recvQueueSize: Long = 128,
    val sendQueueSize: Long = 128,
    val packetPoolSize: Long = 256,
    val batchSize: Long = 32
)

/**
 * VPN connection status
 */
enum class VpnStatus(val value: Int) {
    DISCONNECTED(0),
    CONNECTING(1),
    CONNECTED(2),
    RECONNECTING(3),
    ERROR(4);
    
    companion object {
        fun fromInt(value: Int) = values().firstOrNull { it.value == value } ?: DISCONNECTED
    }
}

/**
 * VPN statistics
 */
data class VpnStats(
    val bytesSent: Long,
    val bytesReceived: Long,
    val packetsSent: Long,
    val packetsReceived: Long,
    val connectedDurationMs: Long,
    val queueDrops: Long,
    val errors: Long
) {
    val connectedDuration: Double
        get() = connectedDurationMs / 1000.0
}

/**
 * Network configuration (from DHCP)
 */
data class NetworkInfo(
    val ipAddress: String,
    val gateway: String,
    val netmask: String,
    val dnsServers: Array<String>,
    val mtu: Int
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        
        other as NetworkInfo
        
        if (ipAddress != other.ipAddress) return false
        if (gateway != other.gateway) return false
        if (netmask != other.netmask) return false
        if (!dnsServers.contentEquals(other.dnsServers)) return false
        if (mtu != other.mtu) return false
        
        return true
    }
    
    override fun hashCode(): Int {
        var result = ipAddress.hashCode()
        result = 31 * result + gateway.hashCode()
        result = 31 * result + netmask.hashCode()
        result = 31 * result + dnsServers.contentHashCode()
        result = 31 * result + mtu
        return result
    }
}

/**
 * VPN exception types
 */
sealed class VpnException(message: String) : Exception(message) {
    class NotCreated : VpnException("VPN client not created")
    class CreateFailed : VpnException("Failed to create VPN client")
    class ConnectFailed(code: Int) : VpnException("Connection failed with code $code")
    class OperationFailed(code: Int) : VpnException("Operation failed with code $code")
    class ReadFailed(code: Int) : VpnException("Read failed with code $code")
    class WriteFailed(code: Int) : VpnException("Write failed with code $code")
    class NoData : VpnException("No data available")
    class Timeout : VpnException("Operation timed out")
    class InvalidNetworkInfo : VpnException("Invalid network information")
}

// ============================================================================
// Main Client Class
// ============================================================================

/**
 * SoftEther Mobile VPN Client
 * 
 * Thread-safe Kotlin wrapper for mobile FFI layer
 */
class SoftEtherMobileClient {
    
    // Native handle
    private val handle = AtomicLong(0)
    
    // State tracking
    private val isInitialized = AtomicBoolean(false)
    
    // Coroutine scope for async operations
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    // Callbacks
    private var statusCallback: ((VpnStatus) -> Unit)? = null
    private var statsCallback: ((VpnStats) -> Unit)? = null
    private var networkCallback: ((NetworkInfo) -> Unit)? = null
    
    companion object {
        private const val TAG = "SoftEtherMobile"
        
        // Load native library
        init {
            try {
                System.loadLibrary("softether_mobile")
            } catch (e: UnsatisfiedLinkError) {
                android.util.Log.e(TAG, "Failed to load native library", e)
                throw e
            }
        }
        
        // Native library info
        external fun nativeGetVersion(): String
        external fun nativeGetBuildInfo(): String
        
        // Library initialization
        private external fun nativeInit(): Int
        private external fun nativeCleanup()
        
        // Initialize library once
        private val initialized = AtomicBoolean(false)
        
        fun ensureInitialized() {
            if (initialized.compareAndSet(false, true)) {
                val result = nativeInit()
                if (result != 0) {
                    throw VpnException.CreateFailed()
                }
            }
        }
        
        /**
         * Get library version
         */
        val version: String
            get() = try {
                nativeGetVersion()
            } catch (e: Exception) {
                "Unknown"
            }
        
        /**
         * Get build info
         */
        val buildInfo: String
            get() = try {
                nativeGetBuildInfo()
            } catch (e: Exception) {
                "Unknown"
            }
    }
    
    // ========================================================================
    // Native Methods
    // ========================================================================
    
    private external fun nativeCreate(
        server: String, port: Int, hub: String, username: String, passwordHash: String,
        useEncrypt: Boolean, useCompress: Boolean, halfConnection: Boolean, maxConnection: Int,
        recvQueueSize: Long, sendQueueSize: Long, packetPoolSize: Long, batchSize: Long
    ): Long
    
    private external fun nativeDestroy(handle: Long)
    private external fun nativeConnect(handle: Long): Int
    private external fun nativeDisconnect(handle: Long): Int
    private external fun nativeGetStatus(handle: Long): Int
    private external fun nativeIsConnected(handle: Long): Boolean
    private external fun nativeReadPacket(handle: Long, buffer: ByteArray, timeoutMs: Int): Int
    private external fun nativeWritePacket(handle: Long, data: ByteArray, length: Int): Int
    private external fun nativeGetStats(handle: Long): VpnStats?
    private external fun nativeGetNetworkInfo(handle: Long): NetworkInfo?
    private external fun nativeGetError(handle: Long): String?
    
    // ========================================================================
    // Public API
    // ========================================================================
    
    /**
     * Create VPN connection with configuration
     */
    fun create(config: VpnConfig) {
        ensureInitialized()
        
        // Clean up existing connection
        destroy()
        
        val h = nativeCreate(
            config.serverName, config.serverPort, config.hubName, 
            config.username, config.password,
            config.useEncrypt, config.useCompress, config.halfConnection, config.maxConnection,
            config.recvQueueSize, config.sendQueueSize, config.packetPoolSize, config.batchSize
        )
        
        if (h == 0L) {
            throw VpnException.CreateFailed()
        }
        
        handle.set(h)
        isInitialized.set(true)
        
        android.util.Log.i(TAG, "VPN client created: handle=$h")
    }
    
    /**
     * Connect to VPN server (blocking)
     */
    fun connect() {
        val h = handle.get()
        if (h == 0L) throw VpnException.NotCreated()
        
        val result = nativeConnect(h)
        if (result != 0) {
            throw VpnException.ConnectFailed(result)
        }
        
        android.util.Log.i(TAG, "VPN connection initiated")
    }
    
    /**
     * Connect to VPN server (async with timeout)
     */
    suspend fun connectAsync(timeoutMs: Long = 30000): Unit = withContext(Dispatchers.IO) {
        connect()
        
        // Wait for connection to establish
        val startTime = System.currentTimeMillis()
        while (System.currentTimeMillis() - startTime < timeoutMs) {
            val status = getStatus()
            
            when (status) {
                VpnStatus.CONNECTED -> {
                    android.util.Log.i(TAG, "VPN connected")
                    return@withContext
                }
                VpnStatus.ERROR -> {
                    throw VpnException.ConnectFailed(-1)
                }
                else -> {
                    delay(100) // Check every 100ms
                }
            }
        }
        
        throw VpnException.Timeout()
    }
    
    /**
     * Disconnect from VPN server (blocking)
     */
    fun disconnect() {
        val h = handle.get()
        if (h == 0L) return
        
        val result = nativeDisconnect(h)
        if (result != 0) {
            android.util.Log.w(TAG, "Disconnect returned: $result")
        }
        
        android.util.Log.i(TAG, "VPN disconnected")
    }
    
    /**
     * Disconnect from VPN server (async)
     */
    suspend fun disconnectAsync(): Unit = withContext(Dispatchers.IO) {
        disconnect()
        
        // Wait for disconnection
        repeat(50) {
            if (getStatus() == VpnStatus.DISCONNECTED) {
                return@withContext
            }
            delay(100)
        }
    }
    
    /**
     * Get current VPN status
     */
    fun getStatus(): VpnStatus {
        val h = handle.get()
        if (h == 0L) return VpnStatus.DISCONNECTED
        
        val statusInt = nativeGetStatus(h)
        return VpnStatus.fromInt(statusInt)
    }
    
    /**
     * Check if connected
     */
    fun isConnected(): Boolean {
        val h = handle.get()
        if (h == 0L) return false
        return nativeIsConnected(h)
    }
    
    /**
     * Get VPN statistics
     */
    fun getStats(): VpnStats {
        val h = handle.get()
        if (h == 0L) throw VpnException.NotCreated()
        
        return nativeGetStats(h) ?: throw VpnException.OperationFailed(-1)
    }
    
    /**
     * Get network configuration
     */
    fun getNetworkInfo(): NetworkInfo {
        val h = handle.get()
        if (h == 0L) throw VpnException.NotCreated()
        
        return nativeGetNetworkInfo(h) ?: throw VpnException.InvalidNetworkInfo()
    }
    
    /**
     * Get last error message
     */
    fun getLastError(): String? {
        val h = handle.get()
        if (h == 0L) return null
        return nativeGetError(h)
    }
    
    // ========================================================================
    // Packet I/O
    // ========================================================================
    
    /**
     * Read packet from VPN (blocking)
     * 
     * @param timeoutMs Timeout in milliseconds (0 = non-blocking)
     * @return Packet data as ByteArray
     * @throws VpnException.NoData if no packet available
     * @throws VpnException.ReadFailed on error
     */
    fun readPacket(timeoutMs: Int = 100): ByteArray {
        val h = handle.get()
        if (h == 0L) throw VpnException.NotCreated()
        
        val buffer = ByteArray(2048)
        val bytesRead = nativeReadPacket(h, buffer, timeoutMs)
        
        return when {
            bytesRead < 0 -> throw VpnException.ReadFailed(bytesRead)
            bytesRead == 0 -> throw VpnException.NoData()
            else -> buffer.copyOf(bytesRead)
        }
    }
    
    /**
     * Read packet from VPN (async)
     */
    suspend fun readPacketAsync(timeoutMs: Int = 100): ByteArray = withContext(Dispatchers.IO) {
        readPacket(timeoutMs)
    }
    
    /**
     * Write packet to VPN (blocking)
     * 
     * @param data Packet data to write
     * @throws VpnException.WriteFailed on error
     */
    fun writePacket(data: ByteArray) {
        val h = handle.get()
        if (h == 0L) throw VpnException.NotCreated()
        
        val result = nativeWritePacket(h, data, data.size)
        if (result != 0) {
            throw VpnException.WriteFailed(result)
        }
    }
    
    /**
     * Write packet to VPN (async)
     */
    suspend fun writePacketAsync(data: ByteArray): Unit = withContext(Dispatchers.IO) {
        writePacket(data)
    }
    
    // ========================================================================
    // Callbacks (Polling-based for now)
    // ========================================================================
    
    /**
     * Set status change callback
     * 
     * Note: Currently uses polling. Native callback support can be added later.
     */
    fun setStatusCallback(callback: (VpnStatus) -> Unit) {
        statusCallback = callback
        
        // Start polling for status changes
        scope.launch {
            var lastStatus = VpnStatus.DISCONNECTED
            while (isInitialized.get()) {
                try {
                    val currentStatus = getStatus()
                    if (currentStatus != lastStatus) {
                        callback(currentStatus)
                        lastStatus = currentStatus
                    }
                } catch (e: Exception) {
                    android.util.Log.e(TAG, "Error in status callback", e)
                }
                delay(500) // Poll every 500ms
            }
        }
    }
    
    /**
     * Set stats update callback
     */
    fun setStatsCallback(callback: (VpnStats) -> Unit) {
        statsCallback = callback
        
        // Start polling for stats
        scope.launch {
            while (isInitialized.get()) {
                try {
                    if (isConnected()) {
                        val stats = getStats()
                        callback(stats)
                    }
                } catch (e: Exception) {
                    android.util.Log.e(TAG, "Error in stats callback", e)
                }
                delay(5000) // Poll every 5 seconds
            }
        }
    }
    
    /**
     * Set network info callback
     */
    fun setNetworkCallback(callback: (NetworkInfo) -> Unit) {
        networkCallback = callback
        
        // Start polling for network info
        scope.launch {
            var lastInfo: NetworkInfo? = null
            while (isInitialized.get()) {
                try {
                    if (isConnected()) {
                        val info = getNetworkInfo()
                        if (info != lastInfo) {
                            callback(info)
                            lastInfo = info
                        }
                    }
                } catch (e: Exception) {
                    // Network info might not be available yet
                }
                delay(2000) // Poll every 2 seconds
            }
        }
    }
    
    // ========================================================================
    // Cleanup
    // ========================================================================
    
    /**
     * Destroy VPN client and free resources
     */
    fun destroy() {
        val h = handle.getAndSet(0)
        if (h != 0L) {
            try {
                disconnect()
            } catch (e: Exception) {
                android.util.Log.e(TAG, "Error disconnecting", e)
            }
            
            nativeDestroy(h)
            android.util.Log.i(TAG, "VPN client destroyed")
        }
        
        isInitialized.set(false)
        
        // Cancel all coroutines
        scope.cancel()
    }
    
    /**
     * Finalize - ensure cleanup
     */
    protected fun finalize() {
        destroy()
    }
}

File: ios/Swift/SoftEtherMobileClient.swift
//
//  SoftEtherMobileClient.swift
//  SoftEtherZig iOS Client
//
//  Swift wrapper for mobile FFI layer (libsoftether_mobile.a)
//

import Foundation
import NetworkExtension

// MARK: - Swift Types

/// VPN connection configuration
public struct VpnConfig {
    public let serverName: String
    public let serverPort: UInt16
    public let hubName: String
    public let username: String
    public let password: String
    
    // Connection options
    public var useEncrypt: Bool = true
    public var useCompress: Bool = true
    public var halfConnection: Bool = false
    public var maxConnection: UInt8 = 1
    
    // Performance tuning
    public var recvQueueSize: UInt64 = 128
    public var sendQueueSize: UInt64 = 128
    public var packetPoolSize: UInt64 = 256
    public var batchSize: UInt64 = 32
    
    public init(serverName: String, serverPort: UInt16, hubName: String, username: String, password: String) {
        self.serverName = serverName
        self.serverPort = serverPort
        self.hubName = hubName
        self.username = username
        self.password = password
    }
}

/// VPN connection status
public enum VpnStatus: Int32 {
    case disconnected = 0
    case connecting = 1
    case connected = 2
    case reconnecting = 3
    case error = 4
    
    var description: String {
        switch self {
        case .disconnected: return "Disconnected"
        case .connecting: return "Connecting"
        case .connected: return "Connected"
        case .reconnecting: return "Reconnecting"
        case .error: return "Error"
        }
    }
}

/// VPN statistics
public struct VpnStats {
    public let bytesSent: UInt64
    public let bytesReceived: UInt64
    public let packetsSent: UInt64
    public let packetsReceived: UInt64
    public let connectedDurationMs: UInt64
    public let queueDrops: UInt64
    public let errors: UInt64
    
    public var connectedDuration: TimeInterval {
        return TimeInterval(connectedDurationMs) / 1000.0
    }
}

/// Network configuration (from DHCP)
public struct NetworkInfo {
    public let ipAddress: String
    public let gateway: String
    public let netmask: String
    public let dnsServers: [String]
    public let mtu: UInt16
    
    init?(fromCStruct cInfo: MobileNetworkInfo) {
        // Convert IP address
        let ip = cInfo.ip_address
        self.ipAddress = "\(ip.0).\(ip.1).\(ip.2).\(ip.3)"
        
        // Convert gateway
        let gw = cInfo.gateway
        self.gateway = "\(gw.0).\(gw.1).\(gw.2).\(gw.3)"
        
        // Convert netmask
        let nm = cInfo.netmask
        self.netmask = "\(nm.0).\(nm.1).\(nm.2).\(nm.3)"
        
        // Convert DNS servers (filter out zeros)
        var dns: [String] = []
        let dnsArray = withUnsafeBytes(of: cInfo.dns_servers) { ptr in
            Array(ptr.bindMemory(to: (UInt8, UInt8, UInt8, UInt8).self))
        }
        for server in dnsArray {
            if server.0 == 0 && server.1 == 0 && server.2 == 0 && server.3 == 0 {
                break
            }
            dns.append("\(server.0).\(server.1).\(server.2).\(server.3)")
        }
        self.dnsServers = dns
        
        self.mtu = cInfo.mtu
    }
}

// MARK: - Main Client Class

/// Swift wrapper for mobile FFI VPN client
public class SoftEtherMobileClient {
    
    // MARK: - Properties
    
    private var handle: MobileVpnHandle?
    private let queue = DispatchQueue(label: "com.softether.mobile.client", qos: .userInitiated)
    
    // Callbacks
    private var statusCallback: ((VpnStatus) -> Void)?
    private var statsCallback: ((VpnStats) -> Void)?
    private var networkCallback: ((NetworkInfo) -> Void)?
    
    // Keep C callback contexts alive
    private var statusContext: UnsafeMutableRawPointer?
    private var statsContext: UnsafeMutableRawPointer?
    private var networkContext: UnsafeMutableRawPointer?
    
    // MARK: - Initialization
    
    public init() {
        // Initialize mobile VPN library
        let result = mobile_vpn_init()
        if result != 0 {
            NSLog("SoftEtherMobile: Warning - init returned \(result)")
        }
    }
    
    deinit {
        disconnect()
        cleanup()
    }
    
    // MARK: - Connection Management
    
    /// Create VPN connection with configuration
    public func create(config: VpnConfig) throws {
        try queue.sync {
            // Clean up existing connection
            if let existingHandle = handle {
                mobile_vpn_destroy(existingHandle)
                handle = nil
            }
            
            // Convert Swift config to C struct
            var cConfig = MobileVpnConfig()
            
            config.serverName.withCString { serverPtr in
                config.hubName.withCString { hubPtr in
                    config.username.withCString { userPtr in
                        config.password.withCString { passPtr in
                            cConfig.server = serverPtr
                            cConfig.port = config.serverPort
                            cConfig.hub = hubPtr
                            cConfig.username = userPtr
                            cConfig.password_hash = passPtr
                            cConfig.use_encrypt = config.useEncrypt
                            cConfig.use_compress = config.useCompress
                            cConfig.half_connection = config.halfConnection
                            cConfig.max_connection = config.maxConnection
                            cConfig.recv_queue_size = config.recvQueueSize
                            cConfig.send_queue_size = config.sendQueueSize
                            cConfig.packet_pool_size = config.packetPoolSize
                            cConfig.batch_size = config.batchSize
                            
                            // Create handle
                            handle = mobile_vpn_create(&cConfig)
                        }
                    }
                }
            }
            
            guard handle != nil else {
                throw VpnError.createFailed
            }
        }
    }
    
    /// Connect to VPN server
    public func connect() throws {
        try queue.sync {
            guard let handle = handle else {
                throw VpnError.notCreated
            }
            
            let result = mobile_vpn_connect(handle)
            if result != 0 {
                throw VpnError.connectFailed(code: result)
            }
        }
    }
    
    /// Disconnect from VPN server
    public func disconnect() {
        queue.sync {
            guard let handle = handle else { return }
            _ = mobile_vpn_disconnect(handle)
        }
    }
    
    /// Async connect with timeout
    public func connect(timeout: TimeInterval = 30.0) async throws {
        try connect()
        
        // Wait for connection to establish
        let startTime = Date()
        while Date().timeIntervalSince(startTime) < timeout {
            let status = getStatus()
            
            if status == .connected {
                return
            }
            
            if status == .error {
                throw VpnError.connectFailed(code: -1)
            }
            
            try await Task.sleep(nanoseconds: 100_000_000) // 100ms
        }
        
        throw VpnError.timeout
    }
    
    /// Async disconnect
    public func disconnect() async {
        disconnect()
        
        // Wait for disconnection
        for _ in 0..<50 {
            if getStatus() == .disconnected {
                return
            }
            try? await Task.sleep(nanoseconds: 100_000_000) // 100ms
        }
    }
    
    // MARK: - Status & Stats
    
    /// Get current VPN status
    public func getStatus() -> VpnStatus {
        return queue.sync {
            guard let handle = handle else { return .disconnected }
            let cStatus = mobile_vpn_get_status(handle)
            return VpnStatus(rawValue: cStatus.rawValue) ?? .disconnected
        }
    }
    
    /// Get VPN statistics
    public func getStats() throws -> VpnStats {
        return try queue.sync {
            guard let handle = handle else {
                throw VpnError.notCreated
            }
            
            var cStats = MobileVpnStats()
            let result = mobile_vpn_get_stats(handle, &cStats)
            
            if result != 0 {
                throw VpnError.operationFailed(code: result)
            }
            
            return VpnStats(
                bytesSent: cStats.bytes_sent,
                bytesReceived: cStats.bytes_received,
                packetsSent: cStats.packets_sent,
                packetsReceived: cStats.packets_received,
                connectedDurationMs: cStats.connected_duration_ms,
                queueDrops: cStats.queue_drops,
                errors: cStats.errors
            )
        }
    }
    
    /// Get network configuration
    public func getNetworkInfo() throws -> NetworkInfo {
        return try queue.sync {
            guard let handle = handle else {
                throw VpnError.notCreated
            }
            
            var cInfo = MobileNetworkInfo()
            let result = mobile_vpn_get_network_info(handle, &cInfo)
            
            if result != 0 {
                throw VpnError.operationFailed(code: result)
            }
            
            guard let info = NetworkInfo(fromCStruct: cInfo) else {
                throw VpnError.invalidNetworkInfo
            }
            
            return info
        }
    }
    
    /// Check if connected
    public func isConnected() -> Bool {
        guard let handle = handle else { return false }
        return mobile_vpn_is_connected(handle)
    }
    
    /// Get last error message
    public func getLastError() -> String? {
        guard let handle = handle else { return nil }
        if let errorPtr = mobile_vpn_get_error(handle) {
            return String(cString: errorPtr)
        }
        return nil
    }
    
    // MARK: - Packet I/O
    
    /// Read packet from VPN (async)
    public func readPacket(timeout: TimeInterval = 0.1) async throws -> Data {
        return try await withCheckedThrowingContinuation { continuation in
            queue.async { [weak self] in
                guard let self = self, let handle = self.handle else {
                    continuation.resume(throwing: VpnError.notCreated)
                    return
                }
                
                // Allocate buffer
                let bufferSize = 2048
                var buffer = [UInt8](repeating: 0, count: bufferSize)
                
                let timeoutMs = UInt32(timeout * 1000)
                let bytesRead = mobile_vpn_read_packet(handle, &buffer, UInt64(bufferSize), timeoutMs)
                
                if bytesRead < 0 {
                    continuation.resume(throwing: VpnError.readFailed(code: Int(bytesRead)))
                } else if bytesRead == 0 {
                    continuation.resume(throwing: VpnError.noData)
                } else {
                    let data = Data(buffer[..<Int(bytesRead)])
                    continuation.resume(returning: data)
                }
            }
        }
    }
    
    /// Write packet to VPN (async)
    public func writePacket(_ data: Data) async throws {
        try await withCheckedThrowingContinuation { continuation in
            queue.async { [weak self] in
                guard let self = self, let handle = self.handle else {
                    continuation.resume(throwing: VpnError.notCreated)
                    return
                }
                
                let result = data.withUnsafeBytes { ptr in
                    mobile_vpn_write_packet(handle, ptr.baseAddress!, UInt64(data.count))
                }
                
                if result != 0 {
                    continuation.resume(throwing: VpnError.writeFailed(code: Int(result)))
                } else {
                    continuation.resume()
                }
            }
        }
    }
    
    /// Synchronous read packet (for blocking operations)
    public func readPacketSync(timeout: TimeInterval = 0.1) throws -> Data {
        return try queue.sync {
            guard let handle = handle else {
                throw VpnError.notCreated
            }
            
            let bufferSize = 2048
            var buffer = [UInt8](repeating: 0, count: bufferSize)
            
            let timeoutMs = UInt32(timeout * 1000)
            let bytesRead = mobile_vpn_read_packet(handle, &buffer, UInt64(bufferSize), timeoutMs)
            
            if bytesRead < 0 {
                throw VpnError.readFailed(code: Int(bytesRead))
            } else if bytesRead == 0 {
                throw VpnError.noData
            }
            
            return Data(buffer[..<Int(bytesRead)])
        }
    }
    
    /// Synchronous write packet
    public func writePacketSync(_ data: Data) throws {
        try queue.sync {
            guard let handle = handle else {
                throw VpnError.notCreated
            }
            
            let result = data.withUnsafeBytes { ptr in
                mobile_vpn_write_packet(handle, ptr.baseAddress!, UInt64(data.count))
            }
            
            if result != 0 {
                throw VpnError.writeFailed(code: Int(result))
            }
        }
    }
    
    // MARK: - Callbacks
    
    /// Set status change callback
    public func onStatusChange(_ callback: @escaping (VpnStatus) -> Void) {
        statusCallback = callback
        
        guard let handle = handle else { return }
        
        // Create context that captures self
        let context = UnsafeMutableRawPointer(Unmanaged.passRetained(self).toOpaque())
        statusContext = context
        
        // Set C callback
        mobile_vpn_set_status_callback(handle, { cStatus, userData in
            guard let userData = userData else { return }
            let client = Unmanaged<SoftEtherMobileClient>.fromOpaque(userData).takeUnretainedValue()
            
            if let status = VpnStatus(rawValue: cStatus.rawValue) {
                client.statusCallback?(status)
            }
        }, context)
    }
    
    /// Set stats update callback
    public func onStatsUpdate(_ callback: @escaping (VpnStats) -> Void) {
        statsCallback = callback
        
        guard let handle = handle else { return }
        
        let context = UnsafeMutableRawPointer(Unmanaged.passRetained(self).toOpaque())
        statsContext = context
        
        mobile_vpn_set_stats_callback(handle, { cStatsPtr, userData in
            guard let userData = userData, let cStatsPtr = cStatsPtr else { return }
            let client = Unmanaged<SoftEtherMobileClient>.fromOpaque(userData).takeUnretainedValue()
            
            let cStats = cStatsPtr.pointee
            let stats = VpnStats(
                bytesSent: cStats.bytes_sent,
                bytesReceived: cStats.bytes_received,
                packetsSent: cStats.packets_sent,
                packetsReceived: cStats.packets_received,
                connectedDurationMs: cStats.connected_duration_ms,
                queueDrops: cStats.queue_drops,
                errors: cStats.errors
            )
            
            client.statsCallback?(stats)
        }, context)
    }
    
    /// Set network info callback
    public func onNetworkInfo(_ callback: @escaping (NetworkInfo) -> Void) {
        networkCallback = callback
        
        guard let handle = handle else { return }
        
        let context = UnsafeMutableRawPointer(Unmanaged.passRetained(self).toOpaque())
        networkContext = context
        
        mobile_vpn_set_network_callback(handle, { cInfoPtr, userData in
            guard let userData = userData, let cInfoPtr = cInfoPtr else { return }
            let client = Unmanaged<SoftEtherMobileClient>.fromOpaque(userData).takeUnretainedValue()
            
            if let info = NetworkInfo(fromCStruct: cInfoPtr.pointee) {
                client.networkCallback?(info)
            }
        }, context)
    }
    
    // MARK: - Cleanup
    
    private func cleanup() {
        if let handle = handle {
            mobile_vpn_destroy(handle)
            self.handle = nil
        }
        
        // Release callback contexts
        if let ctx = statusContext {
            Unmanaged<SoftEtherMobileClient>.fromOpaque(ctx).release()
            statusContext = nil
        }
        if let ctx = statsContext {
            Unmanaged<SoftEtherMobileClient>.fromOpaque(ctx).release()
            statsContext = nil
        }
        if let ctx = networkContext {
            Unmanaged<SoftEtherMobileClient>.fromOpaque(ctx).release()
            networkContext = nil
        }
        
        mobile_vpn_cleanup()
    }
}

// MARK: - Error Types

public enum VpnError: Error, LocalizedError {
    case notCreated
    case createFailed
    case connectFailed(code: Int)
    case operationFailed(code: Int)
    case readFailed(code: Int)
    case writeFailed(code: Int)
    case noData
    case timeout
    case invalidNetworkInfo
    
    public var errorDescription: String? {
        switch self {
        case .notCreated:
            return "VPN client not created"
        case .createFailed:
            return "Failed to create VPN client"
        case .connectFailed(let code):
            return "Connection failed with code \(code)"
        case .operationFailed(let code):
            return "Operation failed with code \(code)"
        case .readFailed(let code):
            return "Read failed with code \(code)"
        case .writeFailed(let code):
            return "Write failed with code \(code)"
        case .noData:
            return "No data available"
        case .timeout:
            return "Operation timed out"
        case .invalidNetworkInfo:
            return "Invalid network information"
        }
    }
}

// MARK: - Version Info

extension SoftEtherMobileClient {
    /// Get library version
    public static var version: String {
        if let versionPtr = mobile_vpn_get_version() {
            return String(cString: versionPtr)
        }
        return "Unknown"
    }
    
    /// Get build info
    public static var buildInfo: String {
        if let buildPtr = mobile_vpn_get_build_info() {
            return String(cString: buildPtr)
        }
        return "Unknown"
    }
}
